Tools/cases_generator/generate_cases.py

```diff
@@ -1,55 +1,326 @@
-arg_parser.add_argument("-q", "--quiet", action="store_true")
+arg_parser.add_argument("-i", "--input", type=str, default=DEFAULT_INPUT)
+arg_parser.add_argument("-o", "--output", type=str, default=DEFAULT_OUTPUT)
 
 
-def eopen(filename: str, mode: str = "r") -> TextIO:
-    if filename == "-":
-        if "r" in mode:
-            return sys.stdin
-        else:
-            return sys.stdout
-    return cast(TextIO, open(filename, mode))
-
-
-def parse_cases(
-    src: str, filename: str|None = None
-) -> tuple[list[InstDef], list[parser.Super], list[parser.Family]]:
-    psr = parser.Parser(src, filename=filename)
-    instrs: list[InstDef] = []
-    supers: list[parser.Super] = []
-    families: list[parser.Family] = []
-    while not psr.eof():
-        if inst := psr.inst_def():
-            instrs.append(inst)
-        elif sup := psr.super_def():
-            supers.append(sup)
-        elif fam := psr.family_def():
-            families.append(fam)
+# This is not a data class
+class Instruction(parser.InstDef):
+    """An instruction with additional data and code."""
+
+    # Computed by constructor
+    always_exits: bool
+    cache_offset: int
+    cache_effects: list[parser.CacheEffect]
+    input_effects: list[parser.StackEffect]
+    output_effects: list[parser.StackEffect]
+
+    # Set later
+    family: parser.Family | None = None
+    predicted: bool = False
+
+    def __init__(self, inst: parser.InstDef):
+        super().__init__(inst.header, inst.block)
+        self.context = inst.context
+        self.always_exits = always_exits(self.block)
+        self.cache_effects = [
+            effect for effect in self.inputs if isinstance(effect, parser.CacheEffect)
+        ]
+        self.cache_offset = sum(c.size for c in self.cache_effects)
+        self.input_effects = [
+            effect for effect in self.inputs if isinstance(effect, parser.StackEffect)
+        ]
+        self.output_effects = self.outputs  # For consistency/completeness
+
+    def write(
+        self, f: typing.TextIO, indent: str, dedent: int = 0
+    ) -> None:
+        """Write one instruction, sans prologue and epilogue."""
+        if dedent < 0:
+            indent += " " * -dedent  # DO WE NEED THIS?
+
+        # Get cache offset and maybe assert that it is correct
+        if family := self.family:
+            if self.name == family.members[0]:
+                if cache_size := family.size:
+                    f.write(
+                        f"{indent}    static_assert({cache_size} == "
+                        f'{self.cache_offset}, "incorrect cache size");\n'
+                    )
+
+        # Write cache effect variable declarations
+        cache_offset = 0
+        for ceffect in self.cache_effects:
+            if ceffect.name != "unused":
+                # TODO: if name is 'descr' use PyObject *descr = read_obj(...)
+                bits = ceffect.size * 16
```

Comment:
I had suggestions on these from the other PR, not sure if you saw (I like `_` more than `unused`, and both `"unused"` and `16` might benefit from constant names in this file).

Unless you just disagreed, in which case, feel free to ignore this. :)